### PR TITLE
ci: add action to enforce adding changelog entries

### PR DESCRIPTION
**Pull request by ewanharris**
## Description

This will run on all PRs and report a failure if they do not modify the changelog, if we do not need a changelog entry for that specific PR we can add a `Skip-Changelog` label on the PR to skip the check. We can customise the label if we don't like that one.

The action used is https://github.com/dangoslen/changelog-enforcer if anyone wants to check out other config we can provide.

Some further N+1 we could make is to either also introduce things such as below, we might need to fork the action to do this:
* Parse the PR title or individual commits and if they follow conventional commits use the commit type as an additional signal for whether the changelog entry is needed (i.e. `docs`, `ci`, `test` shouldn't need an entry, `fix`, `feat`, sometimes `chore` should)
* Introduce an ignore paths field so that PRs that only touch certain folders won't need an entry
* Extend the workflow/action to also run validation if the file was touched

This might land or might not, just wanted to start a discussion about how we improve the changelog process

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: .github/workflows/enforce-changelog-entry.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+on:
+  pull_request:
+    # By default labeled/unlabeled are not included in the pull_request even so we need to list out what we want
+    types: [opened, synchronize, reopened, ready_for_review, labeled, unlabeled]
+
+jobs:
+  changelog:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: dangoslen/changelog-enforcer@v3
```
Comment (ewanharris) [inline .github/workflows/enforce-changelog-entry.yaml:11]: Should be a pinned sha, but I'll do that if we like it

**Review by willvedd — COMMENTED** [review]
Looks like it is doing its job: https://github.com/openfga/openfga/actions/runs/10616883518/job/29428366571?pr=1893

Can you demonstrate applying the skip label?

**Comment by ewanharris** [comment]
@willvedd, voilà!

**Review by miparnisari — APPROVED** [review]

File: .github/workflows/enforce-changelog-entry.yaml (hunk 1 above)
Comment (ewanharris) [reply .github/workflows/enforce-changelog-entry.yaml]: Updated to use a sha so good for merge

File: .github/workflows/enforce-changelog-entry.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+name: "Enforce changelog entry"
```
Comment (willvedd) [inline .github/workflows/enforce-changelog-entry.yaml:1]: **Suggestion:**
```suggestion
name: "Update unreleased section in CHANGELOG"
```
Just to be abundantly clear that this is about the unreleased section and not pertaining to an OpenFGA release.

File: .github/workflows/enforce-changelog-entry.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+on:
+  pull_request:
```
Comment (willvedd) [inline .github/workflows/enforce-changelog-entry.yaml:3]: **Reminder:** We need to skip this for dependabot somehow.

**Comment by willvedd** [comment]
Confirmed that this GH action succeeds when updating the changelog: https://github.com/openfga/openfga/pull/1893/commits/1f6388c39c8cffd91fde3bf56d7cd3f6eacb24f7. However it doesn't validate format, which is fine for now IMO.

File: .github/workflows/enforce-changelog-entry.yaml (hunk 3 above)
Comment (ewanharris) [reply .github/workflows/enforce-changelog-entry.yaml:3]: In 0e9dbf5398365fdb7daece91ba69501ea372375a I extended the labels to now skip on `dependencies` (that is used by dependabot by default) and `tests` (that we use)

**Comment by ewanharris** [comment]
Good callout @willvedd, I added validation into the N+1 list

**Review by willvedd — APPROVED** [review]
Amazing!

**Review by jpadilla — APPROVED** [review]

**Pull request merged**